// @x402/evm/dist/esm/chunk-WKBC5YMI.mjs
import {
  EIP2612_GAS_SPONSORING_KEY,
  ERC20_APPROVAL_GAS_SPONSORING_KEY,
  ERC20_APPROVAL_GAS_SPONSORING_VERSION,
  createPermit2PayloadForProxy
} from "./chunk-CRT6YNY5.mjs";
import {
  DEFAULT_MAX_FEE_PER_GAS,
  DEFAULT_MAX_PRIORITY_FEE_PER_GAS,
  ERC20_APPROVE_GAS_LIMIT,
  PERMIT2_ADDRESS,
  eip2612NoncesAbi,
  eip2612PermitTypes,
  erc20AllowanceAbi,
  erc20ApproveAbi,
  getEvmChainId,
  x402ExactPermit2ProxyAddress
} from "./chunk-C4ZQMS77.mjs";

// src/exact/client/permit2.ts
import { encodeFunctionData, getAddress } from "viem";
var MAX_UINT256 = BigInt("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
async function createPermit2Payload(signer, x402Version, paymentRequirements) {
  return createPermit2PayloadForProxy(
    x402ExactPermit2ProxyAddress,
    signer,
    x402Version,
    paymentRequirements
  );
}
function createPermit2ApprovalTx(tokenAddress) {
  const data = encodeFunctionData({
    abi: erc20ApproveAbi,
    functionName: "approve",
    args: [PERMIT2_ADDRESS, MAX_UINT256]
  });
  return {
    to: getAddress(tokenAddress),
    data
  };
}
function getPermit2AllowanceReadParams(params) {
  return {
    address: getAddress(params.tokenAddress),
    abi: erc20AllowanceAbi,
    functionName: "allowance",
    args: [getAddress(params.ownerAddress), PERMIT2_ADDRESS]
  };
}

// src/shared/extensions.ts
import { getAddress as getAddress4 } from "viem";

// src/exact/client/eip2612.ts
import { getAddress as getAddress2 } from "viem";
async function signEip2612Permit(signer, tokenAddress, tokenName, tokenVersion, chainId, deadline, permittedAmount) {
  const owner = signer.address;
  const spender = getAddress2(PERMIT2_ADDRESS);
  const nonce = await signer.readContract({
    address: tokenAddress,
    abi: eip2612NoncesAbi,
    functionName: "nonces",
    args: [owner]
  });
  const domain = {
    name: tokenName,
    version: tokenVersion,
    chainId,
    verifyingContract: tokenAddress
  };
  const approvalAmount = BigInt(permittedAmount);
  const message = {
    owner,
    spender,
    value: approvalAmount,
    nonce,
    deadline: BigInt(deadline)
  };
  const signature = await signer.signTypedData({
    domain,
    types: eip2612PermitTypes,
    primaryType: "Permit",
    message
  });
  return {
    from: owner,
    asset: tokenAddress,
    spender,
    amount: approvalAmount.toString(),
    nonce: nonce.toString(),
    deadline,
    signature,
    version: "1"
  };
}

// src/exact/client/erc20approval.ts
import { encodeFunctionData as encodeFunctionData2, getAddress as getAddress3, maxUint256 } from "viem";
async function signErc20ApprovalTransaction(signer, tokenAddress, chainId) {
  const from = signer.address;
  const spender = getAddress3(PERMIT2_ADDRESS);
  const data = encodeFunctionData2({
    abi: erc20ApproveAbi,
    functionName: "approve",
    args: [spender, maxUint256]
  });
  const nonce = await signer.getTransactionCount({ address: from });
  let maxFeePerGas;
  let maxPriorityFeePerGas;
  try {
    const fees = await signer.estimateFeesPerGas?.();
    if (!fees) {
      throw new Error("no fee estimates available");
    }
    maxFeePerGas = fees.maxFeePerGas;
    maxPriorityFeePerGas = fees.maxPriorityFeePerGas;
  } catch {
    maxFeePerGas = DEFAULT_MAX_FEE_PER_GAS;
    maxPriorityFeePerGas = DEFAULT_MAX_PRIORITY_FEE_PER_GAS;
  }
  const signedTransaction = await signer.signTransaction({
    to: tokenAddress,
    data,
    nonce,
    gas: ERC20_APPROVE_GAS_LIMIT,
    maxFeePerGas,
    maxPriorityFeePerGas,
    chainId
  });
  return {
    from,
    asset: tokenAddress,
    spender,
    amount: maxUint256.toString(),
    signedTransaction,
    version: ERC20_APPROVAL_GAS_SPONSORING_VERSION
  };
}

// src/shared/rpc.ts
import { createPublicClient, http } from "viem";
var rpcClientCache = /* @__PURE__ */ new Map();
function isConfigByChainId(options) {
  const keys = Object.keys(options);
  return keys.length > 0 && keys.every((key) => /^\d+$/.test(key));
}
function getRpcClient(rpcUrl) {
  const existing = rpcClientCache.get(rpcUrl);
  if (existing) {
    return existing;
  }
  const client = createPublicClient({
    transport: http(rpcUrl)
  });
  rpcClientCache.set(rpcUrl, client);
  return client;
}
function resolveRpcUrl(network, options) {
  if (!options) {
    return void 0;
  }
  if (isConfigByChainId(options)) {
    const chainId = getEvmChainId(network);
    const optionsByChainId = options;
    return optionsByChainId[chainId]?.rpcUrl;
  }
  return options.rpcUrl;
}
function resolveExtensionRpcCapabilities(network, signer, options) {
  const capabilities = {
    signTransaction: signer.signTransaction,
    readContract: signer.readContract,
    getTransactionCount: signer.getTransactionCount,
    estimateFeesPerGas: signer.estimateFeesPerGas
  };
  const needsRpcBackfill = !capabilities.readContract || !capabilities.getTransactionCount || !capabilities.estimateFeesPerGas;
  if (!needsRpcBackfill) {
    return capabilities;
  }
  const rpcUrl = resolveRpcUrl(network, options);
  if (!rpcUrl) {
    return capabilities;
  }
  const rpcClient = getRpcClient(rpcUrl);
  if (!capabilities.readContract) {
    capabilities.readContract = (args) => rpcClient.readContract(args);
  }
  if (!capabilities.getTransactionCount) {
    capabilities.getTransactionCount = async (args) => rpcClient.getTransactionCount({ address: args.address });
  }
  if (!capabilities.estimateFeesPerGas) {
    capabilities.estimateFeesPerGas = async () => rpcClient.estimateFeesPerGas();
  }
  return capabilities;
}

// src/shared/extensions.ts
async function trySignEip2612PermitExtension(signer, options, requirements, result, context) {
  const capabilities = resolveExtensionRpcCapabilities(requirements.network, signer, options);
  if (!capabilities.readContract) {
    return void 0;
  }
  if (!context?.extensions?.[EIP2612_GAS_SPONSORING_KEY]) {
    return void 0;
  }
  const tokenName = requirements.extra?.name;
  const tokenVersion = requirements.extra?.version;
  if (!tokenName || !tokenVersion) {
    return void 0;
  }
  const chainId = getEvmChainId(requirements.network);
  const tokenAddress = getAddress4(requirements.asset);
  try {
    const allowance = await capabilities.readContract({
      address: tokenAddress,
      abi: erc20AllowanceAbi,
      functionName: "allowance",
      args: [signer.address, PERMIT2_ADDRESS]
    });
    if (allowance >= BigInt(requirements.amount)) {
      return void 0;
    }
  } catch {
  }
  const permit2Auth = result.payload?.permit2Authorization;
  const deadline = permit2Auth?.deadline ?? Math.floor(Date.now() / 1e3 + requirements.maxTimeoutSeconds).toString();
  const info = await signEip2612Permit(
    {
      address: signer.address,
      signTypedData: (msg) => signer.signTypedData(msg),
      readContract: capabilities.readContract
    },
    tokenAddress,
    tokenName,
    tokenVersion,
    chainId,
    deadline,
    requirements.amount
  );
  return {
    [EIP2612_GAS_SPONSORING_KEY]: { info }
  };
}
async function trySignErc20ApprovalExtension(signer, options, requirements, context) {
  const capabilities = resolveExtensionRpcCapabilities(requirements.network, signer, options);
  if (!capabilities.readContract) {
    return void 0;
  }
  if (!context?.extensions?.[ERC20_APPROVAL_GAS_SPONSORING_KEY]) {
    return void 0;
  }
  if (!capabilities.signTransaction || !capabilities.getTransactionCount) {
    return void 0;
  }
  const chainId = getEvmChainId(requirements.network);
  const tokenAddress = getAddress4(requirements.asset);
  try {
    const allowance = await capabilities.readContract({
      address: tokenAddress,
      abi: erc20AllowanceAbi,
      functionName: "allowance",
      args: [signer.address, PERMIT2_ADDRESS]
    });
    if (allowance >= BigInt(requirements.amount)) {
      return void 0;
    }
  } catch {
  }
  const info = await signErc20ApprovalTransaction(
    {
      address: signer.address,
      signTransaction: capabilities.signTransaction,
      getTransactionCount: capabilities.getTransactionCount,
      estimateFeesPerGas: capabilities.estimateFeesPerGas
    },
    tokenAddress,
    chainId
  );
  return {
    [ERC20_APPROVAL_GAS_SPONSORING_KEY]: { info }
  };
}

export {
  createPermit2Payload,
  createPermit2ApprovalTx,
  getPermit2AllowanceReadParams,
  trySignEip2612PermitExtension,
  trySignErc20ApprovalExtension
};
